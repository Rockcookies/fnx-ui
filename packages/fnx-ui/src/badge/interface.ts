import { ReactNode, HTMLAttributes } from 'react';

export type BadgeProps = BadgeComponentProps &
	Omit<HTMLAttributes<HTMLSpanElement>, 'color'>;

export interface BadgeComponentProps {
	count?: ReactNode;
	showZero?: boolean;
	dot?: boolean;
	max?: string | number;
	color?: string;
	offset?: [number | string, number | string];
}
