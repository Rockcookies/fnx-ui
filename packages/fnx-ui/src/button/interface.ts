import { ButtonHTMLAttributes, ReactNode } from 'react';

export type ButtonType =
	| 'default'
	| 'primary'
	| 'success'
	| 'warning'
	| 'danger';

export type ButtonSize = 'lg' | 'md' | 'sm' | 'xs';

export type ButtonShape = 'rect' | 'radius' | 'round';

export interface ButtonComponentProps {
	type?: ButtonType;
	size?: ButtonSize;
	icon?: ReactNode;
	iconPosition?: 'left' | 'right';
	loading?: boolean;
	loadingIcon?: ReactNode;
	loadingChildren?: ReactNode;
	disabled?: boolean;
	hairline?: boolean;
	plain?: boolean;
	shape?: ButtonShape;
	block?: boolean;
	color?: string;
	htmlType?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
}

export type ButtonProps = ButtonComponentProps &
	Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'type'>;
