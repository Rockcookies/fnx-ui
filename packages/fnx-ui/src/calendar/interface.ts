import { HTMLAttributes, ReactNode } from 'react';
import CalendarDay from './CalendarDay';
import CalendarMulti from './CalendarMulti';
import CalendarRange from './CalendarRange';

export type CalendarValue = Date | string | number;

export type CalendarDayType =
	| 'normal'
	| 'start'
	| 'start-end'
	| 'middle'
	| 'end'
	| 'selected'
	| 'multiple-middle'
	| 'multiple-selected'
	| 'disabled';

export interface CalendarDayComponentProps {
	color?: string;
	dayHeight?: string;
	date: Date;
	content?: ReactNode;
	type: CalendarDayType;
	topInfo?: ReactNode;
	bottomInfo?: ReactNode;
}

export type CalendarDayProps = CalendarDayComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'content'>;

export interface CalendarSlots {
	day?: (day: CalendarDayComponentProps) => ReactNode;
	subTitle?: boolean | ((date: Date) => ReactNode);
	monthMark?: boolean | ((date: Date) => ReactNode);
	monthTitle?: boolean | ((date: Date) => ReactNode);
	footer?: ReactNode;
}

export interface CalendarBaseProps {
	title?: ReactNode;
	color?: string;
	minDate?: Date | string | number;
	maxDate?: Date | string | number;
	dayHeight?: number | string;
	readonly?: boolean;
	closeIcon?: ReactNode;
	firstDayOfWeek?: number;
	slots?: CalendarSlots;
	confirmText?: ReactNode;
	confirmDisabledText?: ReactNode;
}

export type CalendarElementProps = Omit<
	HTMLAttributes<HTMLDivElement>,
	'defaultValue' | 'onChange' | 'title'
>;

export interface CalendarContainerRef {
	element: HTMLDivElement | null;
	reset: () => void;
	getValue: () => Date[] | undefined;
}

export interface CalendarRef {
	element: HTMLDivElement | null;
	reset: () => void;
	getValue: () => Date | undefined;
}

export interface CalendarComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue;
	onChange?: (value: Date) => void;
	onConfirm?: (value?: Date) => void;
	onCancel?: () => void;
}

export type CalendarProps = CalendarComponentProps & CalendarElementProps;

export type CalendarRangeRef = CalendarContainerRef;

export interface CalendarRangeComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue[];
	onChange?: (value: Date[]) => void;
	onConfirm?: (value?: Date[]) => void;
	onCancel?: () => void;
	allowSameDay?: boolean;
	maxSize?: number;
	onMaxSize?: () => void;
}

export type CalendarRangeProps = CalendarRangeComponentProps &
	CalendarElementProps;

export type CalendarMultiRef = CalendarRangeRef;

export interface CalendarMultiComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue[];
	onChange?: (value: Date[]) => void;
	onConfirm?: (value?: Date[]) => void;
	onCancel?: () => void;
	maxSize?: number;
	onMaxSize?: () => void;
}

export type CalendarMultiProps = CalendarMultiComponentProps &
	CalendarElementProps;

export interface CalendarExportExtra {
	Range: typeof CalendarRange;
	Multi: typeof CalendarMulti;
	Day: typeof CalendarDay;
}
