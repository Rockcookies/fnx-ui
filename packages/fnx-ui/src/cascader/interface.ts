import { HTMLAttributes, ReactNode } from 'react';

export type CascaderValue = string | number;

export interface CascaderOption {
	label?: ReactNode;
	value?: CascaderOption;
	loading?: boolean;
	disabled?: boolean;
	children?: CascaderOption[];
	// for custom filed names
	[key: string]: any;
}

export interface CascaderDataNames {
	label?: string | string[];
	value?: string | string[];
	loading?: string | string[];
	disabled?: string | string[];
	children?: string | string[];
}

export interface CascaderComponentProps<T = CascaderOption> {
	defaultValue?: CascaderValue[];
	data?: T[];
	dataNames?: CascaderDataNames;
	onLoadData?: (options: T[]) => void;
	onConfirm?: (value: CascaderValue[], options: T[]) => void;
	onCancel?: () => void;
	onChange?: (value: CascaderValue[], options: T[]) => void;
	title?: ReactNode;
	placeholder?: ReactNode;
	swipeable?: boolean;
	closeIcon?: ReactNode;
	activeColor?: string;
}

export type CascaderProps<T = CascaderOption> = CascaderComponentProps<T> &
	Omit<HTMLAttributes<HTMLDivElement>, 'defaultValue' | 'title' | 'onChange'>;
