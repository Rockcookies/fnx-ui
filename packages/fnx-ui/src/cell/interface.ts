import { HTMLAttributes, ReactNode } from 'react';
import CellGroup from './CellGroup';

export type CellItemsAlign = 'top' | 'middle' | 'bottom';

export interface CellComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	content?: ReactNode;
	contentProps?: HTMLAttributes<HTMLDivElement>;
	description?: ReactNode;
	descriptionProps?: HTMLAttributes<HTMLDivElement>;
	leftIcon?: ReactNode;
	leftIconProps?: HTMLAttributes<HTMLSpanElement>;
	rightIcon?: ReactNode;
	rightIconProps?: HTMLAttributes<HTMLSpanElement>;
	border?: boolean;
	itemsAlign?: CellItemsAlign;
	clickable?: boolean;
	size?: 'md' | 'lg';
}

export type CellProps = CellComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CellGroupComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	border?: boolean;
	insert?: boolean;
	bodyProps?: HTMLAttributes<HTMLDivElement>;
}

export type CellGroupProps = CellGroupComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CellExportExtra {
	Group: typeof CellGroup;
}
