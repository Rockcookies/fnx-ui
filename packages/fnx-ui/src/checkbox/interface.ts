import { HTMLAttributes, ReactNode } from 'react';
import InternalCheckboxGroup from './CheckboxGroup';

export type CheckboxValue = string | number | boolean;

export interface CheckboxBaseProps {
	disabled?: boolean;
	direction?: 'vertical' | 'horizontal';
	checkedIcon?: ReactNode;
	icon?: ReactNode;
	iconSize?: string | number;
	iconPosition?: 'left' | 'right';
	iconShape?: 'square' | 'round' | 'plain';
	iconCheckedColor?: string;
	labelDisabled?: boolean;
}

export interface CheckboxComponentProps extends CheckboxBaseProps {
	value?: CheckboxValue;
	checked?: boolean;
	defaultChecked?: boolean;
	onChange?: (checked: boolean) => void;
	skipGroup?: boolean;
}

export type CheckboxProps = CheckboxComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'onChange' | 'value'>;

export interface CheckboxRef {
	element: HTMLDivElement | null;
	toggle: () => void;
}

export interface CheckboxGroupComponentProps<
	T extends CheckboxValue = CheckboxValue,
> extends CheckboxBaseProps {
	value?: T[];
	defaultValue?: T[];
	onChange?: (v: T[]) => void;
	maxCheckedCount?: number;
}

export type CheckboxGroupProps<T extends CheckboxValue = CheckboxValue> =
	CheckboxGroupComponentProps<T> &
		Omit<HTMLAttributes<HTMLDivElement>, 'onChange' | 'defaultValue'>;

export interface CheckboxGroupContextData extends CheckboxBaseProps {
	value?: CheckboxValue[];
	registerValue?: (value: CheckboxValue) => void;
	cancelValue?: (value: CheckboxValue) => void;
	toggleValue?: (isCheck: boolean, value: CheckboxValue) => void;
}
export interface CheckboxGroupRef {
	element: HTMLDivElement | null;
	toggleAll: (options?: { checkAll?: boolean }) => void;
}

export interface CheckboxExportExtra {
	Group: typeof InternalCheckboxGroup;
}
