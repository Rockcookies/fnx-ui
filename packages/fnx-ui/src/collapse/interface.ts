import { HTMLAttributes, ReactNode } from 'react';
import { CellProps } from '../cell';
import CollapseItem from './CollapseItem';

export interface CollapseComponentProps {
	accordion?: boolean;
	activeKey?: string | string[];
	defaultActiveKey?: string | string[];
	onChange?: (key: string[]) => void;
	ghost?: boolean;
}

export type CollapseProps = CollapseComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'onChange'>;

export interface CollapseItemComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	headerProps?: CellProps;
	disabled?: boolean;
	ghost?: boolean;
	transitionDuration?: number;
}

export type CollapseItemProps = CollapseItemComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CollapseContextData {
	accordion?: boolean;
	activeKey: string[];
	onChange: (key: string[]) => void;
	ghost: boolean;
}

export interface CollapseExportExtra {
	Item: typeof CollapseItem;
}
