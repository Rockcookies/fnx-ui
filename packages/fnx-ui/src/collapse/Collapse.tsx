import React, {
	cloneElement,
	isValidElement,
	useCallback,
	useMemo,
} from 'react';
import configComponentProps from '../hooks/config-component-props';
import useControlledState from '../hooks/use-controlled-state';
import { classnames, createBEM } from '../utils/namespace';
import { createForwardRef, toElementArray } from '../utils/react';
import CollapseContext from './context';
import { CollapseProps } from './interface';

const NS = 'fnx-collapse';
const bem = createBEM(NS);

const useProps = configComponentProps<
	Required<Pick<CollapseProps, 'defaultActiveKey' | 'accordion' | 'ghost'>>
>({
	defaultActiveKey: [],
	accordion: false,
	ghost: false,
});

const parseKeys = (keys: string | string[]): string[] => {
	const formatted: string[] = [];

	for (const key of Array.isArray(keys) ? keys : [keys]) {
		formatted.push(key);
	}

	return formatted;
};

const Collapse = createForwardRef<HTMLDivElement, CollapseProps>(
	'Collapse',
	(_props, ref) => {
		const [
			{ defaultActiveKey: _defaultActiveKey, accordion, ghost },
			{
				activeKey: __activeKey,
				onChange: _onChange,
				className,
				children,
				...restProps
			},
		] = useProps(_props);

		const _activeKeys = useMemo<string[] | undefined>(
			() => (__activeKey == null ? undefined : parseKeys(__activeKey)),
			[__activeKey],
		);

		const defaultActiveKeys = useMemo<string[]>(
			() => parseKeys(_defaultActiveKey != null ? _defaultActiveKey : []),
			[_defaultActiveKey],
		);

		const { value: activeKey, onChangeRef } = useControlledState({
			value: _activeKeys,
			defaultValue: defaultActiveKeys,
			onChange: _onChange,
		});

		const onChange = useCallback(
			(v: string[]) => onChangeRef.current(v),
			[onChangeRef],
		);

		return (
			<CollapseContext.Provider
				value={{
					activeKey,
					onChange,
					accordion,
					ghost,
				}}
			>
				<div
					className={classnames(bem(), className)}
					{...restProps}
					ref={ref}
				>
					{toElementArray(children).map((child, idx) => {
						if (isValidElement<any>(child)) {
							const key =
								child.key != null ? `${child.key}` : `${idx}`;

							return cloneElement(child, {
								...child.props,
								key,
								'data-fnx-collapse-item-key': key,
							});
						}
					})}
				</div>
			</CollapseContext.Provider>
		);
	},
);

export default Collapse;
