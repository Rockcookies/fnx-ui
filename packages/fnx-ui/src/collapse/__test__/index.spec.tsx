import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import Icon from '../../icon';
import Collapse from '../index';
import { CollapseProps } from '../interface';

describe('<Collapse/>', () => {
	const CollapseDemo = (props: CollapseProps) => {
		return (
			<Collapse {...props}>
				<Collapse.Item key="a" title={<p data-testid="a">a</p>}>
					content
				</Collapse.Item>
				<Collapse.Item key="b" title={<p data-testid="b">b</p>}>
					content
				</Collapse.Item>
				<Collapse.Item key="c" title={<p data-testid="c">c</p>}>
					content
				</Collapse.Item>
			</Collapse>
		);
	};

	it('should update active value when title is clicked', async () => {
		const onChange = jest.fn();

		const { getByTestId } = render(<CollapseDemo onChange={onChange} />);

		fireEvent.click(getByTestId('a'));
		expect(onChange).toBeCalledWith(['a']);

		fireEvent.click(getByTestId('b'));
		expect(onChange).toBeCalledWith(['a', 'b']);

		fireEvent.click(getByTestId('a'));
		expect(onChange).toBeCalledWith(['b']);
	});

	it('should update active value when title is clicked in accordion mode', async () => {
		const onChange = jest.fn();

		const { getByTestId } = render(
			<CollapseDemo onChange={onChange} accordion={true} />,
		);

		fireEvent.click(getByTestId('a'));
		expect(onChange).toBeCalledWith(['a']);

		fireEvent.click(getByTestId('b'));
		expect(onChange).toBeCalledWith(['b']);

		fireEvent.click(getByTestId('b'));
		expect(onChange).toBeCalledWith([]);
	});

	it('should render CollapseItem correctly', async () => {
		const { container } = render(
			<Collapse defaultActiveKey="0">
				<Collapse.Item
					title="title"
					headerProps={{
						content: 'content',
						leftIcon: <Icon name="success" />,
						rightIcon: <Icon name="fail" />,
					}}
				>
					content
				</Collapse.Item>
			</Collapse>,
		);

		expect(container).toMatchSnapshot();
	});
});
