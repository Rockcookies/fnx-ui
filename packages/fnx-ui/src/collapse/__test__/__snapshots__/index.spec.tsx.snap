// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collapse/> should render CollapseItem correctly 1`] = `
<div>
  <div
    class="fnx-collapse"
  >
    <div
      class="fnx-collapse-item fnx-collapse-item--expended"
      data-fnx-collapse-item-key="0"
    >
      <div
        class="fnx-cell fnx-cell--align-top fnx-cell--clickable fnx-collapse-item__header"
        role="button"
        tabindex="0"
      >
        <span
          class="fnx-cell__left-icon"
        >
          <span
            class="fnx-icon"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M122.28096 536.623104c-9.940992-9.925632-11.548672-25.360384-2.78528-36.407296l20.487168-25.828352c8.397824-10.58816 24.108032-13.246464 35.211264-5.835776l177.3312 118.35904c9.353216 6.243328 25.452544 5.430272 34.185216-1.654784l468.5824-380.16c10.532864-8.54528 27.030528-7.817216 36.261888 1.400832l11.542528 11.52512c10.04544 10.03008 9.314304 25.951232-1.215488 36.465664l-502.92736 502.183936c-15.64672 15.624192-41.337856 14.94016-57.445376-1.142784l-219.22816-218.9056z"
              />
            </svg>
          </span>
        </span>
        <div
          class="fnx-cell__title"
        >
          title
        </div>
        <div
          class="fnx-cell__content"
        >
          content
        </div>
        <span
          class="fnx-cell__right-icon"
        >
          <span
            class="fnx-icon fnx-collapse-item__arrow"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M728.223744 520.22784a42.467328 42.467328 0 0 1-11.393024 20.503552L374.90688 882.65728c-16.662528 16.662528-43.677696 16.662528-60.340224 0s-16.662528-43.677696 0-60.340224L626.449408 510.43328 314.614784 198.598656c-16.662528-16.662528-16.662528-43.677696 0-60.340224 16.661504-16.662528 43.676672-16.662528 60.3392 0L716.879872 480.18432c10.860544 10.860544 14.642176 26.120192 11.343872 40.04352z"
              />
            </svg>
          </span>
        </span>
      </div>
      <div
        class="fnx-collapse-item__content-wrapper"
        style="transition-duration: 300ms;"
      >
        <div
          class="fnx-collapse-item__content"
        >
          content
        </div>
      </div>
    </div>
  </div>
</div>
`;
